edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/planner/HealthRestartPlanner.cs

```diff
@@ -135,14 +138,17 @@ public async Task<Plan> CreateShutdownPlanAsync(ModuleSet current)
             IEnumerable<ICommand> dead = await Task.WhenAll(deadTasks);
 
             // create pull, create, update and start commands for added/updated modules
-            IEnumerable<ICommand> addedCommands = await this.ProcessAddedUpdatedModules(
```

Comment:
I think it is cleaner if you create a new command factory for the upfront mode. Older command factory can return NullCommand for PrepareUpdate and no changes to Create or Update commands. Then in health restart planner it can always call PrepareUpdate for upfront, just that if the older command factory was instantiated will return NullCommands.

Reply:
It is possible I am misunderstanding the full picture. But this approach seems not ideal because:
1. What does this save us? IMO that is this logic block. Not huge, but I see where you are coming from. It can be eliminated.
```
                    if (this.moduleUpdateMode == ModuleUpdateMode.NonBlocking)
                    {
                        IList<Task<ICommand>> cmds = new List<Task<ICommand>> { prepareUpdate, createUpdate };
                        // Command needs to be grouped so that image pull is
                        // guaranteed to succeed before we issue a create command.
                        // This prevents multiple create commands from getting
                        // executed within aziot-edged if EdgeAgent timesout
                        // create request (which itself attempts to pull the image)
                        // and reissues.
                        Task<ICommand> createOrUpdateWithUpfrontPull = this.commandFactory.WrapAsync(new GroupCommand(await Task.WhenAll(cmds)));
                        tasks.Add(createOrUpdateWithUpfrontPull);
                    }
                    else
                    {
                        upfrontPullCommands.Add(this.commandFactory.PrepareUpdateAsync(module, runtimeInfo));
                        tasks.Add(createUpdateCommandMaker(moduleWithIdentity));
                    }
```
If true then it becomes this, which is simpler.
```
upfrontPullCommands.Add(this.commandFactory.PrepareUpdateAsync(module, runtimeInfo));
tasks.Add(createUpdateCommandMaker(moduleWithIdentity));
```
3. CommandFactory returns null commands when PrepareUpdateAsync is called, which seems unintuitive. 
4. Lots of same stuff in CommandFactory needs to be duplicated.
5. Having multiple command factories for each type of behavior seems ok with only two types of behaviors. But if we need to change it slightly with another feature flag, will we then need command factory for each combination of feature flags? This will be a lot of duplication if so.

An alternative that seems better approach to me is that we take the problematic code block which contains conditional logic based on feature flag, and move it into helper function (or similar cleanup step). But to keep the spirit of having a command factory that maps to aziot-edged apis so we won't need to have potential combinatorial explosion of command factories if we keep extending. Thoughts?

Reply:
We discussed offline and decided to create an abstraction for a command maker.